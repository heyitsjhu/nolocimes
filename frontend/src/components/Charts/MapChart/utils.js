export const handleOnCountryOver = (event, imageSeries) => {
  rollOverCountry(event.target, imageSeries);
};

export const handleOnCountryOut = (event, imageSeries) => {
  rollOutCountry(event.target, imageSeries);
};

export const handleOnImageOver = (event, polygonSeries) => {
  rollOverCountry(
    polygonSeries.getPolygonById(event.target.dataItem.dataContext.id),
    event.target.parent
  );
};

export const handleOnImageOut = (event, polygonSeries) => {
  rollOutCountry(
    polygonSeries.getPolygonById(event.target.dataItem.dataContext.id),
    event.target.parent
  );
};

// what happens when a country is rolled-over
/**
 *
 * @param {*} mapPolygon
 * @param {*} imageSeries
 */
const rollOverCountry = (mapPolygon, imageSeries) => {
  // resetHover(mapPolygon.parent, imageSeries);

  if (mapPolygon) {
    mapPolygon.isHover = true;

    // make bubble hovered too
    const image = imageSeries.getImageById(mapPolygon.dataItem.dataContext.id);

    if (image) {
      image.dataItem.dataContext.name = mapPolygon.dataItem.dataContext.name;
      image.isHover = true;
      image.dataItem.dataContext.isHover = true;
    }
  }
};

const rollOutCountry = (mapPolygon, imageSeries) => {
  var image = imageSeries.getImageById(mapPolygon.dataItem.dataContext.id);

  resetHover(mapPolygon.parent, imageSeries);
  if (image) {
    image.isHover = false;
  }
};

const resetHover = (mapPolygon, imageSeries) => {
  mapPolygon.mapPolygons.each(polygon => {
    polygon.isHover = false;
  });

  imageSeries.mapImages.each(image => {
    image.isHover = false;
  });
};

/**
 * Positions the tooltip to the top of the circle (bubble) image by
 * offsetting its Y-position by the radius of the circle.
 * @param {*} tooltipY
 * @param {*} target - imageTemplate [amcharts]
 * @returns {number}
 */
export const positionTooltipToTop = (tooltipY, target) => {
  return -target.children.getIndex(0).radius;
};

/**
 *
 * @param {*} positionType
 * @param {*} positionData
 * @param {*} target
 * @param {*} polygonSeries
 * @returns {object}
 */
export const centerImageBubbleOverCountry = (positionType, positionData, target, polygonSeries) => {
  const { dataContext } = target.dataItem;
  var polygon = polygonSeries.getPolygonById(dataContext.id);
  if (polygon) return polygon[`visual${positionType}`];

  return positionData;
};
