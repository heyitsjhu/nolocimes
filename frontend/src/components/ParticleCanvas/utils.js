export const getRandomSpeed = pos => {
  const [min, max] = [-1, 1];
  switch (pos) {
    case 'top':
      return [randomNumFrom(min, max), randomNumFrom(0.1, max)];
    case 'right':
      return [randomNumFrom(min, -0.1), randomNumFrom(min, max)];
    case 'bottom':
      return [randomNumFrom(min, max), randomNumFrom(min, -0.1)];
    case 'left':
      return [randomNumFrom(0.1, max), randomNumFrom(min, max)];
    default:
      return;
  }
};
export const randomArrayItem = arr => {
  return arr[Math.floor(Math.random() * arr.length)];
};
export const randomNumFrom = (min, max) => {
  return Math.random() * (max - min) + min;
};

export const randomSidePos = length => {
  return Math.ceil(Math.random() * length);
};

export const getDisOf = (b1, b2) => {
  const delta_x = Math.abs(b1.x - b2.x);
  const delta_y = Math.abs(b1.y - b2.y);

  return Math.sqrt(delta_x * delta_x + delta_y * delta_y);
};
