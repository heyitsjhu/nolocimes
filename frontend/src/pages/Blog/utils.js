export const stateReducer = (state, action) => {
  if (action.key === 'post') {
    return { ...state, postContent: action.post };
  }

  const { index } = state;
  const prevIndexState = {
    prev: index.prev > 0 ? index.prev - 1 : null,
    current: index.prev,
    next: index.current,
  };
  const nextIndexState = {
    prev: index.current,
    current: index.next,
    next: index.next + 1,
  };
  const indexConfig = action.key === 'prev' ? prevIndexState : nextIndexState;

  return {
    ...state,
    index: indexConfig,
  };
};
