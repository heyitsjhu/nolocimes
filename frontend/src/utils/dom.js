import DLError from 'models/DLError';
import { capitalize } from './';

/**
 *
 * @param {*} type
 * @param {*} name
 */
export const getElClass = (type, name) => {
  if (!name) {
    throw new DLError('Name parameter cannot be empty');
  }

  type = type ? capitalize(type) : '';
  return `DL${type}-${name}`;
};

/**
 *
 * @param {*} type
 * @param {*} id
 */
export const getElId = (type, name) => {
  type = type ? `${type}-` : '';
  return `dlabs-${type}${name || Math.round(Math.random() * 100000)}`;
};

// export const preventDefault = event => event.preventDefault();
