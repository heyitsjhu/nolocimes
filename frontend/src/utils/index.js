export * from './dom';
export * from './formatters';
export * from './postHelpers';
export * from './rfdc';

/**
 *
 * @param {*} str
 */
export const capitalize = str => {
  return str ? str.charAt(0).toUpperCase() + str.slice(1) : '';
};
