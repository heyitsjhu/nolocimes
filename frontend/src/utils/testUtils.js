import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';

import ApplicationProviders from 'store';

const customRender = (ui, options = {}) => {
  const Wrapper = ({ children }) => {
    return (
      <ApplicationProviders>
        <MemoryRouter>{children}</MemoryRouter>
      </ApplicationProviders>
    );
  };

  return render(ui, { wrapper: Wrapper, ...options });
};

export * from '@testing-library/react';
export { customRender as render };
