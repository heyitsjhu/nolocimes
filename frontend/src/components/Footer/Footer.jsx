import React from 'react';
import { useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Link from '@material-ui/core/Link';
import Typography from '@material-ui/core/Typography';
import classnames from 'classnames';

import { CLASSES, ROUTES } from 'const';
import { useCopy } from 'hooks/useCopy';
import { useIsHome } from 'hooks/useIsHome';
import * as Utils from 'utils';

import { GitHubExternalLink, SiteSettingsMenu } from '..';

const useStyles = makeStyles(({ palette, spacing, shared, zIndex }) => ({
  footer: {
    display: 'flex',
    alignItems: 'flex-end',
    justifyContent: 'space-between',
    position: 'fixed',
    bottom: 0,
    left: 0,
    padding: spacing(1),
    width: '100%',
    height: 40,
    opacity: 0,
    zIndex: zIndex.footer,
    [`&.${CLASSES.IS_MOBILE}.${CLASSES.IS_NOT_HOME}`]: {
      backgroundColor: palette.background.default,
    },
  },
  verticalDivider: {
    display: 'inline',
    marginLeft: spacing(1),
    marginRight: spacing(1),
    borderLeft: shared.borderDefault,
    borderColor: palette.text.secondary,
  },
}));

export default () => {
  const { t } = useCopy();
  const history = useHistory();
  const classes = useStyles();
  const siteSettings = useSelector(state => state.siteSettings);
  const isHome = useIsHome();
  const { isOnMobile } = siteSettings;

  return (
    <Box
      id={Utils.getElId('site', 'footer')}
      className={classnames([
        classes.footer,
        !isHome && CLASSES.IS_NOT_HOME,
        isOnMobile && CLASSES.IS_MOBILE,
      ])}
      component="footer"
    >
      <SiteSettingsMenu />
      <Typography color="textSecondary" variant="caption">
        {t('components.Footer.copyright')}
        {!isOnMobile && (
          <>
            <Box className={classes.verticalDivider} />
            <Link color="textSecondary" onClick={() => history.push(ROUTES.PRIVACY_POLICY)}>
              {t('components.Footer.privacyPolicy')}
            </Link>
          </>
        )}
      </Typography>
      <GitHubExternalLink />
    </Box>
  );
};
