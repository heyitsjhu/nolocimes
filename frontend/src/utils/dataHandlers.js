import { DateTime } from 'luxon';
import { deepClone } from 'utils';

export const COVID_CHART_DATA_LABELS = [
  { key: 'cases_new', label: 'New Cases' },
  { key: 'cases_active', label: 'Active Cases' },
  { key: 'cases_critical', label: 'Critical Cases' },
  { key: 'cases_recovered', label: 'Recovered Cases' },
  { key: 'cases_total', label: 'Total Cases' },
  { key: 'deaths_new', label: 'New Deaths' },
  { key: 'deaths_total', label: 'Total Deaths' },
  { key: 'tests_total', label: 'Total Tests' },
];

// Merges Covid-19 historical data for an array of countries into the
// reducer's coronavirus history object for amCharts ingestion. Also, pushes
// the country into a _retrievedCountries property for easier reference
export const convertCovidHistoricalData = (existingData, countriesData) => {
  const newState = existingData ? deepClone(existingData) : {}; // need initial state with keys??
  const stateKeys = Object.keys(existingData).filter(key => key !== '_retrievedCountries');

  countriesData.forEach(countryData => {
    if (countryData.length === 0) return;

    // TODO need a better way to handle updates of differing data points (ie new data points)
    // if (newState._retrievedCountries.includes(countryData[0].country)) return;

    // track which countries we already fetched data for.
    newState._retrievedCountries.push(countryData[0].country);

    countryData.reduce(getEndOfDayStatisticsOnly, []).forEach(dataPoint => {
      const { country, day, time } = dataPoint;

      stateKeys.forEach(stateKey => {
        const keys = stateKey.split(/_(.+)/);
        const value = stateKey === 'population' ? dataPoint[stateKey] : dataPoint[keys[0]][keys[1]];
        const existingDataPointIdx = newState[stateKey].findIndex(d => d.day === dataPoint.day);

        // If given date already has recorded data, add country's data to it; otherwise, create new entry for the current date
        if (existingDataPointIdx > -1) {
          newState[stateKey][existingDataPointIdx][country] = value;
        } else {
          newState[stateKey].push({ [country]: value, day, time });
        }

        // Sort the collection by date (most recent first)
        newState[stateKey].sort((a, b) => DateTime.fromISO(b.day) - DateTime.fromISO(a.day));
      });
    });
  });

  return newState;
};

// Given a country's historical data set, pushes the end of day, or latest, data for each day into, and returns, the accumulated array.
const getEndOfDayStatisticsOnly = (accItem, currItem) => {
  const hasEntry = accItem.find(item => item.day === currItem.day);
  if (!accItem.length || !hasEntry) {
    accItem.push(currItem);
  }

  return accItem;
};

// Converts Covid-19 historical data into a format for amCharts ingestion, based on the
// available, retrieved data
export const convertCovidHistoricalDataForAmcharts = (historicalData = [], countries = []) => {
  const replaceRegex = new RegExp(/-/, 'gi');
  const dataKeys = Object.keys(historicalData).filter(key => key !== '_retrievedCountries');

  return historicalData._retrievedCountries
    .map(country => {
      const countryName = country.replace(replaceRegex, ' ');
      const countryData = countries.find(countryItem => countryItem.name === countryName);
      const countryStatistics = {
        id: countryData ? countryData.alpha2code : undefined,
        name: countryName,
      };

      dataKeys.forEach((dataKey, index) => {
        // zero index gets most recent data point
        // TODO: allow users to adjust a slider to change via day or time
        const data = historicalData[dataKey][0];
        if (index === 0) {
          countryStatistics.day = data.day;
          countryStatistics.time = data.time;
        }

        countryStatistics[dataKey] = data[country];
      });

      return countryStatistics;
    })
    .filter(dataItem => typeof dataItem.id !== 'undefined');
};

// Converts Covid-19 statistics data for amCharts ingestion.
export const convertCovidStatisticsDataForAmcharts = (statisticsData = [], countries = []) => {
  const replaceRegex = new RegExp(/-/, 'gi');

  return statisticsData
    .map(({ country, cases, deaths, tests, day, time }) => {
      const countryName = country.replace(replaceRegex, ' ');
      const countryData = countries.find(countryItem => countryItem.name === countryName);

      return {
        id: countryData ? countryData.alpha2code : undefined,
        name: countryName,
        cases_new: cases.new,
        cases_active: cases.active,
        cases_critical: cases.critical,
        cases_recovered: cases.recovered,
        cases_total: cases.total,
        deaths_new: deaths.new,
        deaths_total: deaths.total,
        tests_total: tests.total,
        day,
        time,
      };
    })
    .filter(dataItem => typeof dataItem.id !== 'undefined');
};
