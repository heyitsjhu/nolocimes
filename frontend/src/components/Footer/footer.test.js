import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { render, fireEvent, screen } from 'utils/testUtils';
import {
  getByLabelText,
  getByText,
  getByTestId,
  queryByTestId,
  waitFor,
} from '@testing-library/dom';

import Footer from './Footer';

jest.mock('const', () => ({ CLASSES: {}, LINKS: {}, POSTS: {}, ROUTES: {}, STORE_KEYS: {} }));

describe('[components] - Footer', () => {
  let container = null;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  test('it renders without failing snapshot', () => {
    const onPrivacyPolicyClickMock = jest.fn();
    const isOpen = true;

    act(() => {
      render(<Footer />);
    });

    fireEvent.click(screen.getByText('Privacy Policy'));

    // expect(screen.getByRole("contentinfo")

    // fireEvent.click(screen.getByRole('presentation'), { button: 1 });
    // expect(onPrivacyPolicyClickMock).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });
});
