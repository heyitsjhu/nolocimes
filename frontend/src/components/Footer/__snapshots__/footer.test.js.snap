// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] - Footer it renders without failing snapshot 1`] = `<div />`;
